### ISSUE TITLE
"GeneFamily with primaryIdentifier 'phytozome_10_2.59256135' not found"

### ISSUE THREAD

**Issue by alancleary** [legumeinfo/graphql-server#36]
I just updated the function that makes GraphQL requests in the `dev/` directory of the web-components repo to throw any errors returned by the GraphQL server and found that apparently some of the gene families with `null` values are actually due to a server error. Here's the full error object that was returned:
```javascript
{
  "message": "GeneFamily with primaryIdentifier 'phytozome_10_2.59256135' not found",
  "locations": [
    {
      "line": 9,
      "column": 35
    }
  ],
  "path": [
    "genes",
    3,
    "geneFamilyAssignments",
    1,
    "geneFamily"
  ],
  "extensions": {
    "code": "BAD_USER_INPUT",
    "stacktrace": [
      "GraphQLError: GeneFamily with primaryIdentifier 'phytozome_10_2.59256135' not found",
      "    at IntermineAPI.inputError (file:///graphql-server/dist/data-sources/intermine/intermine.server.js:46:15)",
      "    at file:///graphql-server/dist/data-sources/intermine/api/get-gene-family.js:12:18",
      "    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"
    ]
  }
}
```
I don't know Intermine super well but it looks like some genes have been assigned gene family identifiers that don't actually exist in the database. If this is indeed an Intermine error, how do you think we should protect against it? We could update the GraphQL resolver to just return null if an error was thrown but this only masks the problem and we would have to check the error logs occasionally to see something's gone wrong. Alternatively, I could update the web-component function that makes GraphQL requests to report the errors in the console, i.e. with `console.error`, but prevent it from throwing an error unless no results were returned. In this case, unless the UI breaks due to incomplete data, someone would have to notice that some data is missing or just happen to see the errors in the console. Neither solution seems great...

The error was in the `add-gene-search` parameters branch. I used the updated gene search web component to search for genes with species "Lotus" as the only constraint.

**Comment by adf-ncgr** [comment]
It seems like this is a multi-faceted issue, mostly having to do with how to handle partial errors in resolution of graphql queries. Just focusing on the "what caused this particular error" aspect of things, I'd guess that although the lotus annotations dataset as a whole has assignments to both the legfed_v1_0 and phytozome_10_2 gene families, there are some genes within it that find matches only to phytozome_10_2 gene families and trigger this issue. Probably the easiest specific fix is not to load the phytozome_10_2 gfa files, and it may be OK to just ignore it for now since it sounds like from the user point of view it behaves just as it would if they weren't loaded? This of course does not answer the more general question of error handling. Of the two choices you listed, I'd probably opt for console logging for now, but also happy to defer to your judgments.

**Comment by sammyjava** [comment]
Actually this is a GraphQL error. I had errantly put GeneFamily.phylotree.primaryIdentifier in the model, but the Phytozome gene families are not associated with a phylotree. Mine queries break when a dependent object is null.

Fixed in commit f91072d

But that's not the point here, the point here is to determine how we deal with GraphQL errors in web-components, but that's a web-components issue, unless we implement something on the GraphQL side. If on the web components side we should move this issue there.

**Comment by sammyjava** [comment]
Yes, that makes sense. The reason I dropped the errors from the API functions was exactly for that reason, to not crash queries that have a null subattribute (which is quite common). Throwing errors in the resolvers allows us to control how that is dealt with in the right place. We should certainly throw them on a get-by-identifier queries. And probably not ever on subattribute resolves.

**Comment by alancleary** [comment]
OK, cool. That should probably be implemented in another branch that we can merge here after it's added to main. Do you want to take care of it or should I do it?

**Comment by sammyjava** [comment]
I'll focus on my add-linkout-service branch for now, but happy to hit this later on if you don't get to it.

**Comment by alancleary** [comment]
The GraphQL part of the solution was implemented in PR #39. [web-components issue 114](https://github.com/legumeinfo/web-components/issues/114) will handle the client side.
